### PR TITLE
add some unit tests for estimateGas

### LINKED ISSUE #323: Setup more sophisticated testing environment in unit test suite
1. With the default `KeeperTestSuite`, `GetCoinbaseAddress` will raise an error, because of proposer address and validators not initialized
2. We need to allocate some funds in genesis.

_Originally posted by @yihuang in https://github.com/tharsis/ethermint/pull/272#discussion_r672757806_

### PR DESCRIPTION
**Pull request by yihuang**
## Description

- add unit tests for estimateGas
- add some test environment setup, Closes #323
- add a precompiled erc20 smart contract in the test case, instruction to compile is included in the comment

______

For contributor use:

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [x] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Comment by yihuang** [comment]
One of the test cases currently fail for estimated gas number different from the one returned by geth.

File: x/evm/keeper/grpc_query.go [hunk 1]
```diff
@@ -406,6 +406,10 @@ func (k Keeper) EstimateGas(c context.Context, req *types.EthCallRequest) (*type
 	ctx := sdk.UnwrapSDKContext(c)
 	k.WithContext(ctx)
 
+	if req.GasCap < ethparams.TxGas {
+		return nil, status.Error(codes.InvalidArgument, "GasCap must not lower than 21000")
```
Comment (fedekunze) [inline x/evm/keeper/grpc_query.go:410]:
```suggestion
		return nil, status.Error(codes.InvalidArgument, "gas cap cannot be lower than 21,000")
```

File: x/evm/keeper/grpc_query_test.go [hunk 2]
```diff
@@ -18,8 +23,27 @@ import (
...
...fffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16815260200190815260200160002054905092915050565b6000828211156040518060400160405280601281526020017f4d4154485f5355425f554e444552464c4f570000000000000000000000000000815250906111ae576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004018080602001828103825283818151815260200191508051906020019080838360005b83811015611173578082015181840152602081019050611158565b50505050905090810190601f1680156111a05780820380516001836020036101000a031916815260200191505b509250505060405180910390fd5b50600082840390508091505092915050565b6000808284019050838110156040518060400160405280601181526020017f4d4154485f4144445f4f564552464c4f57000000000000000000000000000000815250906112a8576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004018080602001828103825283818151815260200191508051906020019080838360005b8381101561126d578082015181840152602081019050611252565b50505050905090810190601f16801561129a5780820380516001836020036101000a031916815260200191505b509250505060405180910390fd5b50809150509291505056fea265627a7a72315820d78a33bfcaf696ab45393e8e57e6cd526c981fcdef7b30c0f5c93baca905bf3764736f6c63430005110032"
```
Comment (fedekunze) [inline x/evm/keeper/grpc_query_test.go:31]: nit, can you move this to another file, eg: `keeper/abi_test.go`, when I open this file it crashes VSCode

**Review by fedekunze — APPROVED** [review]
ACK, one minor suggestion

File: x/evm/keeper/grpc_query_test.go (hunk 2 above)
Comment (yihuang) [reply x/evm/keeper/grpc_query_test.go:31]: sure, maybe both bin and abi, put together into a json file.

Comment (yihuang) [reply x/evm/keeper/grpc_query_test.go:31]: done, use `go:embed`, thus increase golang version to 1.16.

**Comment by yihuang** [comment]
> One of the test cases currently fail for estimated gas number different from the one returned by geth.

the test case failure is due to unclean context, now added capability for test suite to commit and start a new block.

**Comment by fedekunze** [comment]
> the test case failure is due to unclean context, now added capability for test suite to commit and begin a new block.

can you fix the failing test? Then we can merge this PR

**Comment by yihuang** [comment]
> > the test case failure is due to unclean context, now added capability for test suite to commit and begin a new block.
> 
> can you fix the failing test? Then we can merge this PR

that one is already fixed, current failure seems caused by something else, I'm checking it now.

**Comment by yihuang** [comment]
> > the test case failure is due to unclean context, now added capability for test suite to commit and begin a new block.
> 
> can you fix the failing test? Then we can merge this PR

fixed now, remove the commit in test suite setup.

**Pull request merged**